Fix typos and combine unnecessary line breaks (#8010)
jenkins-infra/jenkins.io: content/doc/book/pipeline/jenkinsfile.adoc

}
----

Not all Pipelines will have these same three stages, but it is a good starting point to define them for most projects.
The sections below will demonstrate the creation and execution of a simple Pipeline in a test installation of Jenkins.

[NOTE]
====
It is assumed that there is already a source control repository set up for the project and a Pipeline has been defined in Jenkins following <<getting-started#defining-a-pipeline-in-scm, these instructions>>.
====

Using a text editor, ideally one which supports link:http://groovy-lang.org[Groovy] syntax highlighting, create a new `Jenkinsfile` in the root directory of the project.

The Declarative Pipeline example above contains the minimum necessary structure to implement a continuous delivery pipeline.
The <<syntax#agent, agent directive>>, which is required, instructs Jenkins to allocate an executor and workspace for the Pipeline.
Without an `agent` directive, not only is the Declarative Pipeline not valid, it would not be capable of doing any work!
By default the `agent` directive ensures that the source repository is checked out and made available for steps in the subsequent stages.

> The <<syntax#stages, stages directive>>, and <<syntax#steps, steps directives>> are also required for a valid Declarative Pipeline as they instruct Jenkins what to execute and in which stage it should be executed.

Reviewer: ```suggestion
The <<syntax#stages, stages directive>> and <<syntax#steps, steps directives>> are also required for a valid Declarative Pipeline as they instruct Jenkins what to execute and in which stage it should be executed.
```
Suggestion to remove unneeded comma.

}
// Declarative not yet implemented //
----
<1> The `checkout` step will checkout code from source control; `scm` is a special variable which instructs the `checkout` step to clone the specific revision which triggered this Pipeline run.

=== Build

For many projects the beginning of "work" in the Pipeline would be the "build" stage.
Typically this stage of the Pipeline will be where source code is assembled, compiled, or packaged.

> The `Jenkinsfile` is *not* a replacement for an existing build tool such as GNU/Make, Maven, Gradle, etc, but rather can be viewed as a glue layer to bind the multiple phases of a project's development lifecycle (build, test, deploy, etc) together.

Reviewer: ```suggestion
The `Jenkinsfile` is *not* a replacement for an existing build tool such as GNU/Make, Maven, Gradle, or others, but rather can be viewed as a glue layer to bind the multiple phases of a project's development lifecycle (build, test, deploy) together.
```
Suggestion to remove `etc` as per the [microsoft style guide](https://learn.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/e/etc)

    /* .. snip .. */
}
----
<1> Accessing the `currentBuild.result` variable allows the Pipeline to determine if there were any test failures.
In which case, the value would be `UNSTABLE`.

Assuming everything has executed successfully in the example Jenkins Pipeline, each successful Pipeline run will have associated build artifacts archived, test results reported upon and the full console output all in Jenkins.

> A Scripted Pipeline can include conditional tests (shown above), loops, try/catch/finally blocks and even functions.

Reviewer: ```suggestion
A Scripted Pipeline can include conditional tests (shown above), loops, try/catch/finally blocks, and even functions.
```
Suggestion to add punctuation for Oxford comma.

}
// Script //
----
<1> An `agent` must be set at the top level of the pipeline.
This will fail if agent is set as `agent none`.
<2> When using `returnStdout` a trailing whitespace will be appended to the returned string.
Use `.trim()` to remove this.

=== Handling credentials

Credentials
link:../../using/using-credentials#configuring-credentials[configured in Jenkins] can be handled in Pipelines for immediate use.
Read more about using credentials in Jenkins on the link:../../using/using-credentials[Using credentials] page.

.The correct way to handle credentials in Jenkins
video::yfjtMIDgmfs[youtube,width=800,height=420]

==== For secret text, usernames and passwords, and secret files

Jenkins' declarative Pipeline syntax has the `credentials()` helper method (used within the <<syntax#environment,`environment`>> directive) which supports <<#secret-text,secret text>>, <<#usernames-and-passwords,username and password>>, as well as <<#secret-files,secret file>> credentials.

> If you want to handle other types of credentials, refer to the <<#for-other-credential-types, For other credential types>> section (below).

Reviewer: ```suggestion
If you want to handle other types of credentials, refer to the <<#for-other-credential-types, For other credential types>> section.
```
Suggestion to remove `below` per [microsoft style guide](https://learn.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/b/below)

}
// Script //
----
<1> You can reference the two credential environment variables (defined in this Pipeline's <<syntax#environment,`environment`>> directive), within this stage's steps using the syntax `$AWS_ACCESS_KEY_ID` and `$AWS_SECRET_ACCESS_KEY`.
For example, here you can authenticate to AWS using the secret text credentials assigned to these credential variables.

> To maintain the security and anonymity of these credentials, if the job displays the value of these credential variables from within the Pipeline (e.g. `echo $AWS_SECRET_ACCESS_KEY`), Jenkins only returns the value "`+****+`" to reduce the risk of secret information being disclosed to the console output and any logs.

Reviewer: ```suggestion
To maintain the security and anonymity of these credentials, if the job displays the value of these credential variables from within the Pipeline (such as `echo $AWS_SECRET_ACCESS_KEY`), Jenkins only returns the value "`+****+`" to reduce the risk of secret information being disclosed to the console output and any logs.
```
Suggestion to replace `e.g.` with `such as` per [microsoft style guide](https://learn.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/e/eg)

==== For other credential types

> If you need to set credentials in a Pipeline for anything other than secret text, usernames and passwords, or secret files(<<#for-secret-text-usernames-and-passwords-and-secret-files,above>>) - i.e SSH keys or certificates, then use Jenkins' *Snippet Generator* feature, which you can access through Jenkins' classic UI.

Reviewer: ```suggestion
If you need to set credentials in a Pipeline for anything other than secret text, usernames and passwords, or <<#for-secret-text-usernames-and-passwords-and-secret-files,secret files>> like SSH keys or certificates, use Jenkins' *Snippet Generator* feature, which you can access through Jenkins' classic UI.
```
Suggestion to adjust link to not use [`above`](https://learn.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/a/above) and [replace `i.e` usage ](https://learn.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/i/ie).

To access the *Snippet Generator* for your Pipeline project/item:

> . From the Jenkins home page (i.e. the Dashboard of Jenkins' classic UI), click the name of your Pipeline project/item.

Reviewer: ```suggestion
. From the Jenkins Dashboard, select the name of your Pipeline project/item.
```
Suggestion to update this to be more specific because the Jenkins controller home page is known as the Dashboard in the UI. Also update `click` to be `select` per [microsoft style guide](https://learn.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/c/click)

> . On the left, click *Pipeline Syntax* and ensure that the *Snippet Generator* link is in bold at the top-left. (If not, click its link.)

Reviewer: ```suggestion
. In the the left navigation pane, select *Pipeline Syntax* and ensure that the *Snippet Generator* option is available at the top of the navigation pane.
```
Suggestion to update this text to reflect the current UI better.

===== Combining credentials in one step

Using the *Snippet Generator*, you can make multiple credentials available within a single `withCredentials( ... ) { ... }` step by doing the following:

> . From the Jenkins home page (i.e. the Dashboard of Jenkins' classic UI), click the name of your Pipeline project/item.

Reviewer: ```suggestion
. From the Jenkins Dashboard, select the name of your Pipeline project/item.
```
Suggestion to use Jenkins Dashboard and replace `click` with `select` as per previous comment.

> . On the left, click *Pipeline Syntax* and ensure that the *Snippet Generator* link is in bold at the top-left. (If not, click its link.)

Reviewer: ```suggestion
. In the left navigation pane, select *Pipeline Syntax* and ensure that the *Snippet Generator* option is available at the top of the navigation pane.
```
Suggestion to adjust for current UI, same as above.

. From the *Sample Step* field, choose *withCredentials: Bind credentials to variables*.
. Click *Add* under *Bindings*.
. Choose the credential type to add to the `withCredentials( ... ) { ... }` step from the dropdown list.
. Specify the credential *Bindings* details.
  Read more above these in the procedure under <<#for-other-credential-types,For other credential types>> (above).
. Repeat from "Click *Add* ..." (above) for each (set of) credential/s to add to the `withCredentials( ... ) { ... }` step.

> . Click *Generate Pipeline Script* to generate the final `withCredentials( ... ) { ... }` step snippet.

Reviewer: ```suggestion
. Select *Generate Pipeline Script* to generate the final `withCredentials( ... ) { ... }` step snippet.
```
Suggestion to replace `Click` with `Select`.
